[SwiftUI] Provide ability for clients to enable or disable magnification gestures

rr-codes opened the pull request:
#### 42aebba2a97bea8fe82fb7f77c10008b2097fb3b
<pre>
[SwiftUI] Provide ability for clients to enable or disable magnification gestures
<a href="https://bugs.webkit.org/show_bug.cgi?id=288450">https://bugs.webkit.org/show_bug.cgi?id=288450</a>
<a href="https://rdar.apple.com/145534365">rdar://145534365</a>

Reviewed by Abrar Rahman Protyasha and Tim Horton.

Add a `webViewMagnificationGestures` view modifier to WebView.

* Source/WebKit/_WebKit_SwiftUI/View+WebViewModifiers.swift:
(View.webViewMagnificationGestures(_:)):
* Source/WebKit/_WebKit_SwiftUI/WebView.swift:
* Source/WebKit/_WebKit_SwiftUI/WebViewRepresentable.swift:
(WebViewRepresentable.updatePlatformView(_:context:)):
* Tools/SwiftBrowser/Source/Views/ContentView.swift:
(ContentView.body):

Canonical link: <a href="https://commits.webkit.org/291142@main">https://commits.webkit.org/291142@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e94d98c0182d7d4174c24c0af0242a286a1d075a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92009 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11557 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96932 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42599 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20031 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70592 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28073 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9052 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83331 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50920 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8784 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/935 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41814 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79103 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/928 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98968 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19119 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14142 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79621 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19371 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79186 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78853 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19545 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23382 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/698 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12154 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19100 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24306 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18797 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22255 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20545 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -102,6 +102,15 @@ extension WebPageWebView {
     }
+
+    @_spi(CrossImportOverlay)
+    public var allowsMagnification: Bool {
```
Doesn't this need a macOS implementation?

rr-codes replied:
nope, because macOS WKWebView already has allowsMagnification allowsMagnification

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -102,6 +102,15 @@ extension WebPageWebView {
+
+    @_spi(CrossImportOverlay)
+    public var allowsMagnification: Bool {
+        get { !self._disablesDoubleTapToZoom }
```
Should we instead have an internal `_allowsMagnification` on iOS, and the logic in the setter kept in `WKWebView` itself?

rr-codes replied:
sure, will change

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -429,6 +429,8 @@ struct PerWebProcessState {
 
     RetainPtr<WKScrollGeometry> _currentScrollGeometry;
+
+    BOOL _allowsMagnification;
```
isn't this view level state, not per-WP??

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -984,6 +985,8 @@ FOR_EACH_PRIVATE_WKCONTENTVIEW_ACTION(DECLARE_WKCONTENTVIEW_ACTION_FOR_WEB_VIEW)
 
 - (BOOL)_shouldHideSelectionDuringOverflowScroll:(UIScrollView *)scrollView;
 
+- (void)_updateDoubleTapGestureRecognizers;
```
-RecognizerEnablement?

rr-codes committed:
[SwiftUI] Provide ability for clients to enable or disable magnification gestures
https://bugs.webkit.org/show_bug.cgi?id=288450
rdar://145534365

Reviewed by Abrar Rahman Protyasha and Tim Horton.

Add a `webViewMagnificationGestures` view modifier to WebView.

* Source/WebKit/_WebKit_SwiftUI/View+WebViewModifiers.swift:
(View.webViewMagnificationGestures(_:)):
* Source/WebKit/_WebKit_SwiftUI/WebView.swift:
* Source/WebKit/_WebKit_SwiftUI/WebViewRepresentable.swift:
(WebViewRepresentable.updatePlatformView(_:context:)):
* Tools/SwiftBrowser/Source/Views/ContentView.swift:
(ContentView.body):

Canonical link: https://commits.webkit.org/291142@main
